### PR TITLE
(Dark) Theme implementation

### PR DESCRIPTION
**Pull request by tamnguyen98**
This is a feature I've been working on for my school (Washington State University - Vancouver) version of Autolab. I have communicated with you guys on Slack about this feature and it seems like you guys may be interested in implementing this feature on your repository. You may wonder why there are a lot of changes being erb's file and the reason was that I had to make it work with the main scss stylesheet. 

Summary of this pull request:

- Ability to change the theme from navbar, and account settings
- All visible CSS (like style and navbar) files reference the _theme.css color variables
- All color changes should only need to be made in _theme.css (follow the dark class layout)
- To minimize the number of files to be edited for a future theme, style.css.scss overrides any CSS that is from a third-party tool (such as slickgrid)
    - All tables should have the same (roughly) theme

Minor edits to some layouts consist of:
- Added a divider under the 'Courses' text
- Change the viewFeedback page to output code in a code block
- Fixed some erb's incorrect HTML tags and surrounding text in text tags (e.g., i, p, ref, etc.)
- Remove in-page static style attributes, such as "red" and "white" class (this overrides the theme)

Some things to consider:
- Changing the _theme.css visible variable's name to something meaningful instead of hex values

### CODE REVIEW

**Comment by oliverli** [comment]
Awesome, thanks for the contribution!

Since this is a pretty sizable PR and we're all students approaching finals week, it can take us a little time to review this. Apologies in the advance, and thanks for the patience! We'll definitely keep you in the loop if we need any clarifications, and we'll let you know once we've reviewed or merged this PR!

**Comment by tamnguyen98** [comment]
> Awesome, thanks for the contribution!
> 
> Since this is a pretty sizable PR and we're all students approaching finals week, it can take us a little time to review this. Apologies in the advance, and thanks for the patience! We'll definitely keep you in the loop if we need any clarifications, and we'll let you know once we've reviewed or merged this PR!

No problem, I understand. I still have to figure out why it's failing you guy's CI when it's passing ours.

**Comment by victorhuangwq** [comment]
Hi @tamnguyen98, I have taken an initial look at your PR, and I understand that it is by no means a small feat. There are some key issues that I believe needs to be addressed, however

- The updating of the user schema should be done through active record migration [Active Record Migrations](https://guides.rubyonrails.org/active_record_migrations.html)
- As of now, this PR changes the coloring and style of the "light" mode, and we would like to preserve the original style and colors of Autolab's "light" mode. 
- The dark mode misses out a few areas in the /assessment page.  (Perhaps due to new elements being introduced in the main autolab branch)

Note: We understand that the dark mode doesn't have to be perfect to be merged in, considering that it as it is opt in and in "beta".

**Comment by tamnguyen98** [comment]
> Hi @tamnguyen98, I have taken an initial look at your PR, and I understand that it is by no means a small feat. There are some key issues that I believe needs to be addressed, however
> 
> * The adding of the "theme" column of the user schema should be done through active record migration [Active Record Migrations](https://guides.rubyonrails.org/active_record_migrations.html)
> * As of now, this PR changes the coloring and style of the "light" mode, and we would like to preserve the original style and colors of Autolab's "light" mode.
> * The dark mode misses out a few areas in the /assessment page.  (Perhaps due to new elements being introduced in the main autolab branch)
> 
> Note: We understand that the dark mode doesn't have to be perfect to be merged in, considering that it as it is opt in and in "beta".

Hi Victor!
Thanks for the feedback! To address your concerns (in the same order):
- I went ahead and remove the edit from the schema and used the migration as suggested
- I went back and tried to redo some of the theme colors and style to match the original as possible, though not all page make be exactly same due to the reusing the theme variables (there are too many different shades of the same colors in the original style! lol)
- That is correct, and I went ahead and fixed it! but some pages may not have proper theming features since there is some page that I'm unaware of to test! So, I hope more pages will be fixed with the contributions from other members!

Let me know if there are any other concerns I can maybe address!

**Comment by victorhuangwq** [comment]
Okay, we will take a look at your new commits and get back to you.

**Comment by victorhuangwq** [comment]
Hi @tamnguyen98, I have reviewed your code further, and I think there are still some aspects of it in the light mode that needs changing, but it is much better than before. Here are the next steps we need to take to merge this into master.

1. Address the remaining of the issues on light mode
2. Have a better way of managing the variable names. You have mentioned this step in your initial comment, and I think this is very important even now, as I am afraid it can get a lot messier in the future.

I do believe that you have done a lot of the initial setup required for this theme change, but it is still not ready to be merged into master as of the current state. 

However, I would like to help with some of the changes as well. So I have changed the merge destination to a new branch in the project https://github.com/autolab/Autolab/tree/themechooser, that way our team can also make commits to it as well.

**Comment by victorhuangwq** [comment]
@tamnguyen98 Actually, I realized that you as the pull request creator have granted permission for us to make commits into your fork directly. If that was intentional (or okay with you), we can make changes to your fork directly instead. Because in that way you do not need to create a PR for us to review every time you commit.

File: db/migrate/20200516021201_add_theme_to_users.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class AddThemeToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :theme, :string
+    change_column_default :users, :theme, "default"
```
Comment (victorhuangwq) [inline db/migrate/20200516021201_add_theme_to_users.rb:4]: Should use from to, else the change is not reversible
https://guides.rubyonrails.org/active_record_migrations.html#changing-columns

**Comment by tamnguyen98** [comment]
> @tamnguyen98 Actually, I realized that you as the pull request creator have granted permission for us to make commits into your fork directly. If that was intentional (or okay with you), I would prefer that we make changes to your fork directly instead. Because in that way you do not need to create a PR for us to review every time you commit.

Sorry for the late reply! I've been busy doing job hunting and interviews since my graduation. I wouldn't mind at all granting you guys permission, I'm just not sure how (if it hasn't been done already)

**Comment by victorhuangwq** [comment]
@tamnguyen98 no worries! Thank you for making time out for this PR. You have already granted permission for us to push to your fork! I am just checking with you to make sure you did that intentionally / okay with that :). The team and I will make our commits to your branch then, and we will continue to communicate with you through this comment thread.

**Comment by damianhxy** [comment]
Briefly closing and then reopening this to refresh CI actions and also prevent accidental merging.

**Comment by damianhxy** [comment]
Closing this Stale PR. It is still available via issue #1260 if someone plans to pick this up in the future.

**Pull request closed without merging**